28857 openapi include exclude

benjamin-confino opened the pull request:
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

resolves #28857

This PR 
- Creates a centralized service for openAPI to read the relevent parts of the server.xml
- Uses that service to read what applications/modules are included or excluded, overriding mpConfig if both are in use.

benjamin-confino committed:
Use config from mpOpenAPI if available

jantley-ibm commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -23,4 +23,10 @@ docPath=MicroProfile OpenAPI document endpoint path
+uiPath.desc=Specifies the URL path for accessing the MicroProfile OpenAPI UI. The URL path must be constructed with Unicode alphanumeric characters A-Za-z0-9, underscore (_), dash (-), forward slash (/), or period (.) If no value is specified, the path is set to $docPath/ui.
+
+include=MicroProfile OpenAPI included modules.
+include=A comma separated list of modules that will be included in the OpenAPI document. Modules are in the format of {ApplicationName}/{ModuleName} or {ApplicationName} which refer to to <TODO, sort out the confusion of names>. An ApplicationName which does not specify a module will include all modules in the application. The special value "all" will include all available modules.
```
Following appears to be a note that further work needs to be done:
to to <TODO, sort out the confusion of names>

Same appears in line 32

Change
comma separated
to
comma-separated

Change 
that will be included
to
that are to be included

Change
An ApplicationName which
to
An ApplicationName that

Change
will include all modules
to
includes all modules

Change
The special value "all" will include
to
The "all" special value includes

jantley-ibm commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -23,4 +23,10 @@ docPath=MicroProfile OpenAPI document endpoint path
+include=A comma separated list of modules that will be included in the OpenAPI document. Modules are in the format of {ApplicationName}/{ModuleName} or {ApplicationName} which refer to to <TODO, sort out the confusion of names>. An ApplicationName which does not specify a module will include all modules in the application. The special value "all" will include all available modules.
+
+exclude=MicroProfile OpenAPI excluded modules.
+exclude=A comma separated list of that will be excluded from the OpenAPI document. Modules are in the format of {ApplicationName}/{ModuleName} or {ApplicationName} which refer to to <TODO, sort out the confusion of names>. An ApplicationName which does not specify a module will exclude all modules in the application. An exclude takes precedent over an include.
```
Following appears to be a note that further work needs to be done:
to to <TODO, sort out the confusion of names>

Change
A comma separated list of that will be excluded
to
A comma-separated list of modules that are excluded

Change
An exclude takes precedent over an include.
to 
Exclusion supersedes inclusion.

jantley-ibm reviewed (commented):
ID review comments attached; there appears to be a placeholder reminder in a couple of spots, so when that is completed I will review that text.
If you have questions or concerns about any of the suggested changes, please let me know.
I will add the "ID reviewed" label when we agree on changes. Thanks!

benjamin-confino commented:
#build

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -23,72 +23,136 @@
+        }
 
         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-            Tr.debug(null, tc, "Names in config: " + config.getPropertyNames());
+            if (serverxmlmode) {
+                Tr.debug(null, tc, "Acquired includes statement from server.xml");
+            } else {
+                Tr.debug(null, tc, "Names in config: " + configFromMPConfig.getPropertyNames());
+            }
             Tr.debug(null, tc, "Inclusion read from config: " + inclusion);
         }
```
When these debug calls moved from a static method to a non-static method, you need to add the instance.

`Tr.debug(null, tc, ...)` becomes `Tr.debug(this, tc, ...)`

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -23,72 +23,136 @@
 
-        return result;
+        hasBeenInit = true;
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
+            Tr.exit(tc, "lazyInit", toString());
+        }
```
```suggestion
        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
            Tr.exit(this, tc, "lazyInit", toString());
        }
```

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -173,16 +243,55 @@ public List<String> findIncludesNotMatchingAnything(Collection<? extends ModuleI
+        return matches(name, module, serverxmlmode);
+    }
+
+    private boolean matches(ModuleName name, ModuleInfo module, boolean overridenServerXMLMode) {
+        lazyInit();
         if (name.moduleName != null && !name.moduleName.equals(module.getName())) {
             return false;
         }
```
Should this method be static? Does it need to `lazyInit`?

benjamin-confino replied:
Its a private method that is only called from methods that already lazyInit, so we could shave a nanosecond or two by removing that

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -173,16 +243,55 @@ public List<String> findIncludesNotMatchingAnything(Collection<? extends ModuleI
+     */
     private boolean matches(ModuleName name, ModuleInfo module) {
+        return matches(name, module, serverxmlmode);
+    }
```
Does this method need to call `lazyInit` before using `serverxmlmode`?

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/merge/ModelEquality.java:
```diff
@@ -49,14 +49,24 @@ private static boolean equalsImpl(Object a, Object b) {
 
+        /*
+         * Take https://github.com/smallrye/smallrye-open-api/blob/3.13.0/core/src/main/java/io/smallrye/openapi/api/models/ExtensibleImpl.java
+         * extensions is null. If you call addExtension() it will become an map with an item.
+         * if you then call remove, it will remove the item from the list and not null it.
+         * In short new ExtensibleImpl().addExtension(x).removeExtension(x) is not equals() new ExtensibleImpl().
+         *
+         * So to work around that, we treat null and empty maps as equivalent.
+         *
+         * This may need to be extended to cover lists
+         */
         if (a == null) {
-            if (b == null) {
+            if (nullOrEmptyMap(b)) {
                 return true;
             } else {
                 return false;
             }
         } else if (b == null) {
-            return false;
+            return nullOrEmptyMap(a);
         }
```
I think this is ok, but it's close. The [Security Requirement Object][sro] cares about a map being empty vs not being present in an array, but that doesn't quite hit this condition and I'm not aware of other cases.

[sro]: https://spec.openapis.org/oas/v3.1.0.html#security-requirement-object

benjamin-confino replied:
I added this because of one the FAT tests failed, I'd have to take this out to remind myself which on but from memory this lack of equality was causing us to output something like "merged document" when the fat test expected "Generated API".

Let me know if you'd like us to do a deep dive into this?

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/utils/OpenAPIUtils.java:
```diff
@@ -207,6 +210,9 @@ public static <T> boolean allEqual(Collection<? extends T> collection, BiPredica
      * @return {@code true} if {@code a} and {@code b} are equal, {@code false} otherwise
      */
     public static <T> boolean equals(T a, T b, BiPredicate<? super T, ? super T> comparator) {
+
+        System.out.println("GREPGREP " + a + " " + b + " " + comparator.getClass().toGenericString());
+
```
Remove this.

Azquelt replied:
Remove this (or make it proper entry trace)

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/MergeServerXMLTest.java:
```diff
@@ -0,0 +1,510 @@
+
+    @BeforeClass
+    public static void setupServer() throws Exception {
+        server.swapInServerXMLFromPublish(SERVER_XML_NAME);
```
Please use `LibertyServer.updateServerConfiguration(ServerConfiguration)` here instead, so that it's easy to see what the config is in the test class, rather than going to look at an external file.

It also means we don't have two server.xml files to keep in sync.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -17,6 +17,10 @@
+        <AD id="includeApp" name="%includeApp" description="%include.desc" required="false" type="String" cardinality="40000" ibm:beta="true"/>
+        <AD id="excludeApp" name="%excludeApp" description="%exclude.desc" required="false" type="String" cardinality="40000" ibm:beta="true"/>
+        <AD id="includeModule" name="%includeModule" description="%include.desc" required="false" type="String" cardinality="40000" ibm:beta="true"/>
+        <AD id="excludeModule" name="%excludeModule" description="%exclude.desc" required="false" type="String" cardinality="40000" ibm:beta="true"/>
```
```suggestion
        <AD id="includeApplication" name="%includeApplication" description="%includeApplication.desc" required="false" type="String" cardinality="40000" ibm:beta="true"/>
        <AD id="excludeApplication" name="%excludeApplication" description="%excludeApplication.desc" required="false" type="String" cardinality="40000" ibm:beta="true"/>
        <AD id="includeModule" name="%includeModule" description="%includeModule.desc" required="false" type="String" cardinality="40000" ibm:beta="true"/>
        <AD id="excludeModule" name="%excludeModule" description="%excludeModule.desc" required="false" type="String" cardinality="40000" ibm:beta="true"/>
```

Same change needs to be made throughout the implementation and tests (sorry, I know the names of these elements was one of the last things to be decided).

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -23,4 +23,16 @@ docPath=MicroProfile OpenAPI document endpoint path
\ No newline at end of file
+uiPath.desc=Specifies the URL path for accessing the MicroProfile OpenAPI UI. The URL path must be constructed with Unicode alphanumeric characters A-Za-z0-9, underscore (_), dash (-), forward slash (/), or period (.) If no value is specified, the path is set to $docPath/ui.
+
+includeApp=MicroProfile OpenAPI included application.
+includeApp=An application, one per element, that are to be included in the OpenAPI document. The application name is as defined in the server.xml, with a default name of the archive filename excluding any extension.  The "all" special value includes all available applications. 
+
+excludeApp=MicroProfile OpenAPI excluded modules.
+excludeApp=An application, one per element, that are to be included in the OpenAPI document. The application name is as defined in the server.xml, with a default name of the archive filename excluding any extension.  The "all" special value excludes all applications.
+
+includeModule=MicroProfile OpenAPI included modules.
+includeModule=A module, one per element, that are to be included in the OpenAPI document. Module names are in the format {ApplicationName}/{ModuleName}. ModuleName is the Jakarta EE module name and is logged under the message SRVE0169I. 
+
+excludeModule=MicroProfile OpenAPI excluded modules.
+excludeModule=A  module, one per element, that are to be included in the OpenAPI document. Module names are in the format {ApplicationName}/{ModuleName}. ModuleName is the Jakarta EE module name and is logged under the message SRVE0169I.
```
```suggestion
includeApplication=MicroProfile OpenAPI included applications
includeApplication.desc=Application names, one per element, that are to be included in the OpenAPI document. The name of an application may be defined in the server.xml, otherwise it defaults to the application archive filename excluding any extension.  The "all" special value includes all available applications. 

excludeApplication=MicroProfile OpenAPI excluded applications
excludeApplication.desc=Application names, one per element, that are to be excluded from the OpenAPI document.

includeModule=MicroProfile OpenAPI included modules
includeModule.desc=Module names, one per element, that are to be included in the OpenAPI document. Module names must be provided in the format {ApplicationName}/{ModuleName}. The name of a module may be defined in its deployment descriptor, otherwise it defaults to the module archive filename excluding any extension. It is also logged in the SRVE0169I message.

excludeModule=MicroProfile OpenAPI excluded modules
excludeModule.desc=Module names, one per element, that are to be excluded from the OpenAPI document.
```

The keys in this file must match the metatype.xml.

I've reworded the messages a bit and fixed some cases where it said included vs. excluded, though I'm sure ID will have some more suggestions.

I'm not sure whether we should add to this:
- default values
- the versions of mpOpenAPI that this config is valid for
- how the configuration elements interact with each other
  - in theory, the description of these parameters should be sufficient to determine all the test cases required

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,15 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
 OPEN_API_PATH_SEGMENT_INVALID_CWWKO1677E.explanation=OpenAPI path segment "/." or "/.." is invalid.
 OPEN_API_PATH_SEGMENT_INVALID_CWWKO1677E.useraction=Remove any path "/." or "/.." segment from the path.
+
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W=CWWKO1678W: Path segment "/" is in {0} invalid for applications.
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.explanation=explanation: An application cannot name cannot include a "/" as "/" is used to separate the application from its component modules.
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.useraction=useraction: Remove the "/" from the includeApplication or excludeApplication element in your server.xml, or convert that element to includeModule or excludeModule. 
+
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W=CWWKO1679W: Path segment "/" is required for modules but was absent in {0}.
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W.explanation=explanation: An module must be in the format <application-name>/<module-name>, therefore a module name without a "/" is invalid.
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W.useraction=useraction: Correct the includeModule or excludeModule element in your server.xml, or convert that element to includeApplication or excludeApplication. 
```
Message guidelines say that that "unrecoverable configuration errors" are errors while "configuration issues" are warnings.

My initial instinct was that these should be errors, but reading the guidelines I think you're correct that they're warnings because they don't prevent the server from running.

Azquelt replied:
```suggestion
# {1} - the server.xml element name, either includeApplication or excludeApplication
OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W=CWWKO1678W: The configured {0} application name in the {1} element is not valid because it contains the "/" character.
OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.explanation=explanation: An application cannot name cannot include a "/" as "/" is used to separate the application from its component modules.
OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.useraction=useraction: Remove the "/" from the includeApplication or excludeApplication element in your server.xml, or convert that element to includeModule or excludeModule to reference a single module.

# {1} - the server.xml element name, either includeModule or excludeModule
OPEN_API_SLASH_IN_MODULE_CWWKO1679W=CWWKO1679W: The configured {0} module name in the {1} element is not valid it is not in the form "<application-name>/<module-name>".
OPEN_API_SLASH_IN_MODULE_CWWKO1679W.explanation=explanation: An module must be in the format <application-name>/<module-name>, therefore a module name without a "/" is invalid.
OPEN_API_SLASH_IN_MODULE_CWWKO1679W.useraction=useraction: Correct the includeModule or excludeModule element in your server.xml, or convert that element to includeApplication or excludeApplication to reference a whole application.
```

My suggestion adds the element name as a second parameter which would need to be provided in the code and tested.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,15 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+
+OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W=CWWKO1680W: The {0} configuration property includes "{1}". That does not match any deployed application by name but it would match an application by deployment description.
+OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W.explanation=The server is configured to create OpenAPI documentation for the application or module name, this name matches a deployment descriptor but not the application name as defined in server.xml.
+OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W.useraction=Check that all applications started correctly and that the names in the configuration property are specified correctly.
```
It's a bit annoying, but we can't refer to server.xml elements as a "configuration property". I think this message only applies to server.xml elements, but `CWWKO1667W` would apply to both?

We either need to make `CWWKO1667W` more general, or have separate messages for MP Config vs. server.xml.

Azquelt replied:
```suggestion
OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W=CWWKO1680W: The {1} application name in the {0} configuration element does not match the name of any deployed application but it does match the name from the deployment descriptor of the {2} application. The application name used here must be the application name specified in server.xml, or the archive file name with the extension removed if no name is specified in server.xml.
OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W.explanation=The application name reference in the mpOpenAPI configuration element matches the name from an application deployment descriptor, but does not match any application name defined in server.xml.
OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W.useraction=Check that all applications started correctly and that the names in the configuration element are specified correctly.
```

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,189 @@
+    private static final String INVALID_MODULE_WARNING = "OPEN_API_SLASH_IN_MODULE_CWWKO1679W";
+
+    private static final String INCLUDE_APP_PROPERTY_NAME = "includeApp";
+    private static final String EXCLUDE_APP_PROPERTY_NAME = "excludeApp";
```
```suggestion
    private static final String INCLUDE_APP_PROPERTY_NAME = "includeApplication";
    private static final String EXCLUDE_APP_PROPERTY_NAME = "excludeApplication";
```

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,189 @@
+    protected void activate(BundleContext context, Map<String, Object> properties) {
+        if (ProductInfo.getBetaEdition()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(tc, "Initial processing of server.xml");
```
```suggestion
                Tr.event(this, tc, "Initial processing of server.xml");
```

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,189 @@
+    protected void modified(Map<String, Object> properties) {
+        if (ProductInfo.getBetaEdition()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(tc, "Processing update to server.xml");
```
```suggestion
                Tr.event(this, tc, "Processing update to server.xml");
```

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,189 @@
+    protected void deactivate() {
+        if (ProductInfo.getBetaEdition()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(tc, "Deactivating OpenAPIAppConfigProviderImpl");
```
```suggestion
                Tr.event(this, tc, "Deactivating OpenAPIAppConfigProviderImpl");
```

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,189 @@
+        Optional<String> includedModulesString = includedModulesArray.map(this::moduleStringProcessing);
+        Optional<String> excludedModulesString = excludedModulesArray.map(this::moduleStringProcessing);
+
+        //The surprisingly complex code to combine two optionals, excluding null or blank strings
+        includedString = Stream.of(includedAppString,
+                                   includedModulesString).filter(Optional::isPresent).map(Optional::get).filter(StringUtils::isNotBlank).reduce((first, second) -> first + ','
+                                                                                                                                                                   + second);
+
+        excludedString = Stream.of(excludedAppString,
+                                   excludedModulesString).filter(Optional::isPresent).map(Optional::get).filter(StringUtils::isNotBlank).reduce((first, second) -> first + ','
+                                                                                                                                                                   + second);
```
I dislike converting an array of Strings into a comma separated list here when we're going to convert it back into a list later.

Azquelt commented on dev/com.ibm.ws.cdi.api_fat/autoFVT.zip:
Remove this.

Azquelt commented on dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java:
I would suggest moving this change into a separate PR as it's the only change in fattest.simplicity.

Changing any of the core test framework projects will mean your personal builds need to run every bucket, rather than only the ones which use your feature and therefore take a while longer to complete.

I wouldn't shy away from making changes to the test framework if they're needed for your PR, but if it's just doc changes like this I'd move it to its own PR to avoid the larger PB.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/services/OpenAPIAppConfigProvider.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    Optional<String> getExcludedModules();
+
+    public static interface OpenAPIAppConfigListener extends Comparable<OpenAPIAppConfigListener> {
+        public void processConfigUpdate();
+
+        public int getConfigListenerPriority();
+    }
```
This interface needs documented.

If it implements `Comparable`, it needs to also document how two objects are compared and whether that comparison is consistent with `equals`. However, in this case I think it would make more sense **not** to make this implement `Comparable`, and instead use a `Comparator` which compares based on the value returned by `getConfigListenerPriority()`.

In addition, please document whether listeners are called with the largest priority first or the smallest priority first.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/bnd.bnd:
```diff
@@ -1,5 +1,5 @@
 #*******************************************************************************
-# Copyright (c) 2020, 2024 IBM Corporation and others.
+# Copyright (c) 2020, 2022 IBM Corporation and others.
```
Incorrect copyright date change?

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ApplicationRegistryImpl.java:
```diff
@@ -66,13 +67,33 @@ public class ApplicationRegistryImpl implements ApplicationRegistry {
     @Reference
     private MergeProcessor mergeProcessor;
 
+    @SuppressWarnings("unused")
+    private OpenAPIAppConfigProvider openAPIAppConfigProvider;
```
Remove `SuppressWarnings`

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ApplicationRegistryImpl.java:
```diff
@@ -66,13 +67,33 @@ public class ApplicationRegistryImpl implements ApplicationRegistry {
+    @SuppressWarnings("unused")
+    private OpenAPIAppConfigProvider openAPIAppConfigProvider;
+
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.STATIC, unbind = "unbindAppConfigListener")
```
`unbind` is not necessary here, it gets inferred based on the method names.

Azquelt replied:
We agreed this can be left as-is.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ApplicationRegistryImpl.java:
```diff
@@ -66,13 +67,33 @@ public class ApplicationRegistryImpl implements ApplicationRegistry {
+    private OpenAPIAppConfigProvider openAPIAppConfigProvider;
+
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.STATIC, unbind = "unbindAppConfigListener")
+    public void bindAppConfigListener(OpenAPIAppConfigProvider openAPIAppConfigProvider) {
+        this.openAPIAppConfigProvider = openAPIAppConfigProvider;
+        openAPIAppConfigProvider.registerAppConfigListener(this);
+    }
+
+    public void unbindAppConfigListener(OpenAPIAppConfigProvider openAPIAppConfigProvider) {
+        openAPIAppConfigProvider.unregisterAppConfigListener(this);
+        this.openAPIAppConfigProvider = null;
+    }
+
+    @Reference
+    public void bindOpenAPIAppConfigProvider(OpenAPIAppConfigProvider openAPIAppConfigProvider) {
+        this.openAPIAppConfigProvider = null;
+        openAPIAppConfigProvider.registerAppConfigListener(this);
+    }
```
This looks like you have two references to `OpenAPIAppConfigProvider`, but the first one is named `appConfigListener`.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -23,72 +23,136 @@
 
+    private OpenAPIAppConfigProvider configFromServerXML;
+
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.STATIC, unbind = "unbindAppConfigListener")
+    public void bindAppConfigListener(OpenAPIAppConfigProvider openAPIAppConfigProvider) {
+        this.configFromServerXML = openAPIAppConfigProvider;
+        openAPIAppConfigProvider.registerAppConfigListener(this);
+    }
+
+    public void unbindAppConfigListener(OpenAPIAppConfigProvider openAPIAppConfigProvider) {
+        openAPIAppConfigProvider.unregisterAppConfigListener(this);
+        this.configFromServerXML = null;
+    }
```
Since this is a static reference, it might be easier to annotate the field and register as a listener in activate/deactivate.

Technically here we're registering as a listener before we're activated.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -218,6 +327,17 @@ public String toString() {
         }
     }
 
+    @Override
+    public synchronized void processConfigUpdate() {
+        //Reset everything for the next lazyInit
+        hasBeenInit = false;
+        isAll = false;
+        isFirst = false;
+        serverxmlmode = false;
+        included = null;
+        excluded = null;
+    }
```
We have an issue with thread safety because we have this method to reset everything and re-initialize on next use.

If `lazyInit` was only called once it would be sufficient for only the `lazyInit` method to be synchronized as long as every method which needs to access any of these values calls `lazyInit` before doing so.

However, since we have the ability to reset, every use of any of these variables must be synchronized, otherwise they could be reset in the middle of a method which uses them.

To avoid this, I would suggest:
- add a new static inner class to hold these values - let's call it `ConfigValues`
- rather than having a `lazyInit` method which sets all these fields, have a method which creates, caches and returns a `ConfigValues` object
  - all other code which needs the `ConfigValues` object must retrieve it via this method
- `processConfigUpdate` then only needs to delete the cached `ConfigValues` object
- this method and the method to get `ConfigValues` may need synchronization or correct use of `volatile`, but everything else should be correct without synchronization

benjamin-confino replied:
All sounds good, though I think the inner field should not be static since every ConfigValue object will be associated with an outer ModuleSelectionConfigImpl object that it will occasionally have to reference to access the OpenAPIAppConfigProvider.

Azquelt replied:
I was imagining that the outer class would extract the config and pass all those values into the inner class so it wouldn't need to  access the `OpenAPIAppConfigProvider`, but making the inner class non-static so it can access that class itself also seems fine.

However, an important point is that the inner object is immutable so that its values won't change once a method has retrieved it.

If the config gets updated, we would create a new instance of the inner object and future calls would get the new object.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -23,72 +23,136 @@
+        }
     }
 
     @Override
     public String toString() {
+        lazyInit();
         StringBuilder sb = new StringBuilder("Module Selection Config[");
         if (isFirst) {
```
I'm a bit concerned about `toString()` doing initialization here. That's the sort of thing that can cause enabling trace to change behaviour...

Could we make it so that `toString()` reports "Uninitialized Config` or something similar if initialization hasn't happened yet?

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -173,16 +243,55 @@ public List<String> findIncludesNotMatchingAnything(Collection<? extends ModuleI
 
-        return includedNotYetSeen.stream().map(ModuleName::toString).collect(toList());
+        List<ModuleName> includedNotYetSeenButSeenUnderOldNaming = new ArrayList<>();
+        if (serverxmlmode && ProductInfo.getBetaEdition()) {
+            for (Iterator<ModuleName> iterator = includedNotYetSeen.iterator(); iterator.hasNext();) {
+                ModuleName moduleName = iterator.next();
+                for (ModuleInfo moduleInfo : moduleInfos) {
+                    if (matches(moduleName, moduleInfo, false)) {
+                        includedNotYetSeenButSeenUnderOldNaming.add(moduleName);
+                        iterator.remove();
+                        break;
+                    }
+                }
+            }
+
+            for (String unmatchedInclude : includedNotYetSeenButSeenUnderOldNaming.stream().map(ModuleName::toString).collect(toList())) {
+                Tr.warning(tc, MessageConstants.OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W, Constants.MERGE_INCLUDE_CONFIG, unmatchedInclude);
+            }
+        }
```
Can we do this check that the may have used the wrong name for excluded apps as well as included apps please?

(The other check for a name that's missing should remain as only checking the included apps).

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/utils/MessageConstants.java:
```diff
@@ -34,6 +34,7 @@ public class MessageConstants {
     public static final String OPENAPI_MERGE_INFO_PARSE_ERROR_CWWKO1665W = "OPENAPI_MERGE_INFO_PARSE_ERROR_CWWKO1665W"; //$NON-NLS-1$
     public static final String OPENAPI_MERGE_INVALID_NAME_CWWKO1666W = "OPENAPI_MERGE_INVALID_NAME_CWWKO1666W"; //$NON-NLS-1$
     public static final String OPENAPI_MERGE_UNUSED_INCLUDE_CWWKO1667W = "OPENAPI_MERGE_UNUSED_INCLUDE_CWWKO1667W"; //$NON-NLS-1$
+    public static final String OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W = "OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W";
```
There were three messages added but only one message constant. Are the other messages not used yet?

benjamin-confino replied:
The other messages are called from OpenAPIAppConfigProviderImpl.java, so I couldn't store them in this file without creating a circular dependency

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ApplicationRegistryImpl.java:
```diff
@@ -282,6 +288,18 @@ private OpenAPIProvider getFirstProvider() {
 
+    @Override
+    public synchronized void processConfigUpdate() {
+        Map<String, ApplicationRecord> oldApps = applications;
+        applications = new LinkedHashMap<>();
+        for (ApplicationRecord record : oldApps.values()) {
+            //Add application uses config to decide if it creates and registers any providers in ApplicationInfo
+            //Rather than map from the old state to the new state when the config changes, KISS and start again.
+            addApplication(record.info);
+        }
+        cachedProvider = null;
```
In general, we use `synchronized (this)` in the method body rather than `synchronized` in the signature so that the generated trace is outside the synchronized block, making it easier to debug lock contention.

Since `addApplication` can take a while to complete, I think we should do that here.

Azquelt replied:
This has not been done.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ApplicationRegistryImpl.java:
```diff
@@ -291,4 +309,18 @@ public ApplicationRecord(ApplicationInfo info) {
+    }
+
+    @Override
+    public int getConfigListenerPriority() {
+        return 2;
+    }
```
Comment that this is to ensure we're called after the `ModuleSelectionConfig`

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,189 @@
+        }
+    }
+
+    @Modified
+    protected void modified(Map<String, Object> properties) {
+        if (ProductInfo.getBetaEdition()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(tc, "Processing update to server.xml");
+            }
+            processModuleConfig(properties);
+            for (OpenAPIAppConfigListener listener : openAPIAppConfigListeners) {
+                listener.processConfigUpdate();
+            }
+        }
+    }
```
This class only cares about a subset of properties from the `<mpOpenAPI>` config element. If someone changes a config value that isn't related to module inclusion/exclusion then we shouldn't call the listeners since re-generating all the OpenAPI documents is a relatively expensive operation.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/test1/web.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Used for tests that need a non-default deployment name for a war -->
+<web-app version="3.0"
+        xmlns="http://java.sun.com/xml/ns/javaee"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://java.sun.com/xml/ns/javaee web-app_3_0.xsd"
+        id="testID">
+        <module-name>nameFromWar</module-name>
+</web-app>
```
Please use `WebXmlAsset` instead of having a web.xml file included manually.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/MergeServerXMLTestWithUpdate.java:
```diff
@@ -0,0 +1,267 @@
+        assertRest("/test3/test");
+        //Swap the server.xml from saying everything is excluded to saying its included
+        logOnServer("/test1/log", "testTwoWarsUpdate2");
+        server.swapInServerXMLFromPublish(SERVER_XML_NAME);
+        server.waitForStringInLogUsingMark("CWWKG0017I"); //The server configuration was successfully updated
```
```suggestion
        server.setMarkToEndOfLog();
        server.swapInServerXMLFromPublish(SERVER_XML_NAME);
        server.waitForConfigUpdateInLogUsingMark(null);
```
I think we need to manually set the mark ourselves before starting the update if we want to use it afterwards.

Also there's a specific method for waiting for config updates.

But also, I would prefer it if we had a method to update the config that made it more obvious in the test what the value of the config is so you don't have to jump between lots of files to understand the test.

E.g. something that let you make calls like this:
```java
updateConfig(server, includeApp("all"));
updateConfig(server, includeApp("app1"));
updateConfig(server,
    includeApp("app1"),
    includeApp("app2"),
    excludeModule("app2/war1"));
```

or like this:
```java
openapiConfig(server).includeApp("all").apply();
openapiConfig(server).includeApp("app1").apply();
openapiConfig(server).includeApp("app1")
    .includeApp("app2")
    .excludeModule("app2/war1")
    .apply();
```

**Edit:** I've since commented on `MergeServerXmlTest` that we should probably cover the same cases that are covered in `MergeConfigTest`. Those tests could be conducted using dynamic updates, though I'm unsure whether it would cover the same cases you have in this class where there are apps deployed and you update the config.

I wanted to update this message though in case you were working through and implemented this suggestion before seeing the other comment.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/MergeServerXMLTestWithUpdate.java:
```diff
@@ -0,0 +1,267 @@
+
+        //Swap the server.xml from saying everything is excluded to saying its included
+        logOnServer("/test1/log", "testTwoWarsUpdate1");
+        server.swapInServerXMLFromPublish(APP1_XML_NAME);
+        server.waitForStringInLogUsingMark("CWWKG0017I"); //The server configuration was successfully updated
```
```suggestion
        server.setMarkToEndOfLog();
        server.swapInServerXMLFromPublish(APP1_XML_NAME);
        server.waitForConfigUpdateInLogUsingMark(null);
```

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/publish/servers/OpenAPIMergeTestServer/jvm.options:
```diff
@@ -0,0 +1 @@
+-Dcom.ibm.ws.beta.edition=false
```
Do we need this? It should be false by default, right?

benjamin-confino replied:
It put in in there because we're explicitly depending the value being false but its not actually needed.

Azquelt replied:
Ok, makes sense.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/StartupWarningMessagesTest.java:
```diff
@@ -0,0 +1,73 @@
+                                            "CWWKO1680W"),
+                   hasSize(1));
+
+    }
```
This test method contains lots of things are aren't relevant and hides or doesn't explain everything important about the test.

I don't think it's relevant that the app is an ear, or that it contains three modules, or that one of the modules contains a web.xml. Please simplify this.

I'm unsure if the `mp_openapi_extensions_liberty_merged_include` property is important - this looks to be identical to the config in the server.xml? Should it be removed?

The server.xml gives a custom name to the application and adds configuration which includes the app using the wrong name. These details should either be explained in a comment or be moved into the test method.

The error message appears because the server.xml name is "nameFromServerXml" while the deployment descriptor name is "testEar" because there's no deployment descriptor so it defaults to the archive name without the extension. This detail should be explained in a comment.

There's no description of what the `CWWKO1680W` message is or why we expect it.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/StartupWarningMessagesTest.java:
```diff
@@ -0,0 +1,73 @@
+                                             Collections.singletonMap("mp_openapi_extensions_liberty_merged_include", "testEar"));
+        server.startServer();
+
+        //Test that a warning message saying we cant find the app but could under the pre OpenAPI 4.0 rules 
+        assertThat(
+                   server.findStringsInLogs(
+                                            "CWWKO1680W"),
+                   hasSize(1));
```
The warning message is wrong, the rules differ depending on whether the config came from server.xml or MP Config, not whether we're using `mpOpenAPI-4.0` or an earlier feature.

We need to ensure the message has been substituted correctly as the name of the message also includes `CWWKO1680W`. Since this message has placeholders, we should include the expected values for those placeholders in the regex.

I'm not sure if this message is guaranteed to have been written by the time that `server.startServer()` returns. We should use `server.waitForStringInLogs` here unless we're certain that waiting is never necessary.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/StartupWarningMessagesTest.java:
```diff
@@ -0,0 +1,73 @@
+                                    .addClasses(DeploymentTestApp.class, DeploymentTestResource.class)
+                                    .addAsWebInfResource(DeploymentTestApp.class.getPackage(), "web.xml", "web.xml");
+
+        ear = ShrinkWrap.create(EnterpriseArchive.class, "testEar.ear")
+                        .addAsModules(war1, war2, war3);
+        ShrinkHelper.exportAppToServer(server, ear, SERVER_ONLY, DISABLE_VALIDATION);
```
We're disabling validation here because the name of the app doesn't match the archive name. That's probably worth a comment.

Since we do expect the app to start, we should do the validation manually by calling `server.addInstalledAppForValidation("nameFromServerXml");` ourselves after exporting the app.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/MergeServerXMLTest.java:
```diff
@@ -0,0 +1,510 @@
+ * <p>
+ * These tests supplement the MergeProcessor unit tests
+ */
+@RunWith(FATRunner.class)
+public class MergeServerXMLTest {
```
This test class appears to have been adapted from `MergeTest`?

`MergeTest` is meant to test that we correctly merge documents, all tests run with a simple config which merges all apps.

I don't think we need to repeat the same tests with with the same basic config in server.xml.

`MergeConfigTest` is our main test for the merging config - we should include all the include and exclude cases from there in our server.xml config testing. Since the server.xml config is dynamic, we won't have to restart the server between each test like we do in `MergeConfigTest`.

The new cases that you've covered in this class where you're updating the config are obviously still valuable.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ApplicationProcessor.java:
```diff
@@ -48,6 +48,7 @@
 import io.openliberty.microprofile.openapi20.internal.cache.ConfigSerializer;
 import io.openliberty.microprofile.openapi20.internal.services.ConfigFieldProvider;
 import io.openliberty.microprofile.openapi20.internal.services.ModelGenerator;
+import io.openliberty.microprofile.openapi20.internal.services.ModuleSelectionConfig;
```
spurious import

benjamin-confino replied:
Previously ApplicationProcessor referenced ModuleSelectionConfig by being in the same package as it. But ModuleSelectionConfig became ModuleSelectionConfigImpl (its now declarative service) with an interface in a different package. So the import is needed.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -173,16 +243,55 @@ public List<String> findIncludesNotMatchingAnything(Collection<? extends ModuleI
+
+            for (String unmatchedInclude : includedNotYetSeenButSeenUnderOldNaming.stream().map(ModuleName::toString).collect(toList())) {
+                Tr.warning(tc, MessageConstants.OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W, Constants.MERGE_INCLUDE_CONFIG, unmatchedInclude);
+            }
```
This code needs updating because the placeholders in the message have been updated.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,17 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
 OPEN_API_PATH_SEGMENT_INVALID_CWWKO1677E.explanation=OpenAPI path segment "/." or "/.." is invalid.
 OPEN_API_PATH_SEGMENT_INVALID_CWWKO1677E.useraction=Remove any path "/." or "/.." segment from the path.
+
+# {1} - the server.xml element name, either includeApplication or excludeApplication
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W=CWWKO1678W: The configured {0} application name in the {1} element is not valid because it contains the "/" character.
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.explanation=explanation: An application cannot name cannot include a "/" as "/" is used to separate the application from its component modules.
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.useraction=useraction: Remove the "/" from the includeApplication or excludeApplication element in your server.xml, or convert that element to includeModule or excludeModule to reference a single module.
+
+# {1} - the server.xml element name, either includeModule or excludeModule
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W=CWWKO1679W: The configured {0} module name in the {1} element is not valid it is not in the form "<application-name>/<module-name>".
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W.explanation=explanation: An module must be in the format <application-name>/<module-name>, therefore a module name without a "/" is invalid.
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W.useraction=useraction: Correct the includeModule or excludeModule element in your server.xml, or convert that element to includeApplication or excludeApplication to reference a whole application.
```
```suggestion
# {1} - the server.xml element name, either includeApplication or excludeApplication
OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W=CWWKO1678W: The configured {0} application name in the {1} element is not valid because it contains the "/" character.
OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.explanation=An application cannot name cannot include a "/" as "/" is used to separate the application from its component modules.
OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.useraction=Remove the "/" from the includeApplication or excludeApplication element in your server.xml, or convert that element to includeModule or excludeModule to reference a single module.

# {1} - the server.xml element name, either includeModule or excludeModule
OPEN_API_SLASH_IN_MODULE_CWWKO1679W=CWWKO1679W: The configured {0} module name in the {1} element is not valid it is not in the form "<application-name>/<module-name>".
OPEN_API_SLASH_IN_MODULE_CWWKO1679W.explanation=An module must be in the format <application-name>/<module-name>, therefore a module name without a "/" is invalid.
OPEN_API_SLASH_IN_MODULE_CWWKO1679W.useraction=Correct the includeModule or excludeModule element in your server.xml, or convert that element to includeApplication or excludeApplication to reference a whole application.
```

Azquelt commented on dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/MpOpenAPIElement.java:
```diff
@@ -137,7 +137,12 @@ private MpOpenAPIElementBuilder(LibertyServer server) throws CloneNotSupportedEx
+                server.waitForStringInLogUsingMark("CWWKG0017I"); //The server configuration was successfully updated
+                //Setting a low timeout because this only appears if OpenAPI trace is enabled (and enabling trace breaks debuggers)
+                server.waitForStringInLogUsingMark("Finished creating OpenAPI provider", 500);
+            }
```
You need to set the mark before updating the configuration.

Azquelt commented on dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/MpOpenAPIElement.java:
```diff
@@ -46,9 +66,153 @@ public void setUiPath(String uiPath) {
+            this.element = serverConfig.getMpOpenAPIElement();
+        }
+
+        public void buildAndPushToServer() throws Exception {
+            LOG.info("Pushing new server configuration: " + serverConfig.toString());
+            server.updateServerConfiguration(serverConfig);
+            if (server.isStarted()) {
+                server.setMarkToEndOfLog();
+                server.waitForStringInLogUsingMark("CWWKG0017I"); //The server configuration was successfully updated
+                //Setting a low timeout because this only appears if OpenAPI trace is enabled (and enabling trace breaks debuggers)
+                server.waitForStringInLogUsingMark("Finished creating OpenAPI provider", 500);
+            }
+        }
```
I don't really like having this here. In general the classes within this package are concerned with modelling the configuration and not with reading or writing the config to the server.

Let's leave it for now because I want to get this merged and revise it in a future PR.

Azquelt commented on dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java:
```diff
@@ -7102,6 +7102,14 @@ public void deleteAllDropinApplications() throws Exception {
         LibertyFileManager.createRemoteFile(machine, getServerRoot() + "/dropins").mkdir();
     }
 
+    /**
+     * Deletes the folder <serverRoot/apps> and all its contents, then creates a new empty directory in in its place
+     */
+    public void deleteAllAppDirectoryApplications() throws Exception {
+        LibertyFileManager.deleteLibertyDirectoryAndContents(machine, getServerRoot() + "/dropins");
+        LibertyFileManager.createRemoteFile(machine, getServerRoot() + "/dropins").mkdir();
+    }
```
This is wrong - the method says it's deleting from /apps but it's actually deleting dropins. Since it doesn't work, presumably we're not using it anywhere and can remove it?

benjamin-confino replied:
We can remove it, I think it was used in a previous iteration but I left it there for completeness since there's a method adding to apps.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/StartupWarningMessagesTest.java:
```diff
@@ -0,0 +1,83 @@
+        List<String> messages = new ArrayList<>();
+        messages.add("CWWKO1680W.*testEarIncluded.*includeApplication or includeModule.*serverXMLNameIncluded");
+        messages.add("CWWKO1680W.*testEarExcluded.*excludeApplication or excludeModule.*serverXMLNameExcluded");
+        server.waitForStringsInLogUsingMark(messages);
```
Add a comment that this method does actually assert that the messages appear (unlike most of the others which don't).

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/StartupWarningMessagesTest.java:
```diff
@@ -0,0 +1,83 @@
+        server.startServer();
+
+        //Example message: CWWKO1680W: The testEar application name in the includeApplication or includeModule configuration element does not match the name of any deployed application but it does match the name from the deployment descriptor of the serverXMLName application. The application name used here must be the application name specified in server.xml, or the archive file name with the ext ...
+        List<String> messages = new ArrayList<>();
+        messages.add("CWWKO1680W.*testEarIncluded.*includeApplication or includeModule.*serverXMLNameIncluded");
+        messages.add("CWWKO1680W.*testEarExcluded.*excludeApplication or excludeModule.*serverXMLNameExcluded");
+        server.waitForStringsInLogUsingMark(messages);
```
Can we add a third app to test where its server.xml name matches the deployment descriptor name from another app? Config for that name should not produce the warning (since the name does match a deployed application).

benjamin-confino replied:
Sure

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,249 @@
+
+import io.openliberty.microprofile.openapi.internal.common.services.OpenAPIAppConfigProvider;
+
+@Component(configurationPolicy = ConfigurationPolicy.OPTIONAL, immediate = true, configurationPid = "io.openliberty.microprofile.openapi")
```
I don't think this meeds to be immediate?

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,455 @@
+     * @param module the module to check
+     * @return {@code true} if the module should be used, {@code false} otherwise
+     */
+    @Override
+    public boolean isIncluded(ModuleInfo module) {
+        if (getConfigValues().isFirst) {
+            return true;
+        }
+
+        boolean result = false;
+        if (getConfigValues().isAll) {
+            result = true;
+        } else {
+            for (ModuleName name : getConfigValues().included) {
+                if (matches(name, module)) {
+                    result = true;
+                    break;
+                }
+            }
+        }
+
+        if (result) {
+            for (ModuleName name : getConfigValues().excluded) {
+                if (matches(name, module)) {
+                    result = false;
+                    break;
+                }
+            }
+        }
+
+        return result;
+    }
```
We have a problem with consistency in many of the methods in this class.

Most methods rely on having a consistent view of the config that doesn't change throughout the method. In that case, they need to call `getConfigValues()` only once and re-use the result.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,249 @@
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Deprecated
+    public Optional<List<String>> getIncludedAppsAndModules() {
+        betaFenceCheck();
+        List<String> combinedList = new ArrayList<String>(includedApps);
+        combinedList.addAll(includedModules);
+        return combinedList.isEmpty() ? Optional.<List<String>> empty() : Optional.of(Collections.unmodifiableList(combinedList));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Deprecated
+    public Optional<List<String>> getExcludedAppsAndModules() {
+        betaFenceCheck();
+        List<String> combinedList = new ArrayList<String>(excludedApps);
+        combinedList.addAll(excludedModules);
+        return combinedList.isEmpty() ? Optional.<List<String>> empty() : Optional.of(Collections.unmodifiableList(combinedList));
+    }
+
+    @Override
+    @Deprecated
+    public Optional<ConfigMode> getConfigMode() {
+        betaFenceCheck();
+        return configMode;
+    }
```
We have a thread safety issue here (and in other places in the class).

When the config is modified, fields `includedApps` and `includedModules` get updated. If they're modified while we're copying their values (in the `ArrayList` constructor or the `addAll` method), we could get an exception.

Secondly, because there are three separate methods here, there's no way for the caller to get a consistent view of the config. The values could be updated between a call to the first method and a call to the second.

As you suggested, we could fix this with an inner class which holds a snapshot of the config and a method which returns that.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,249 @@
+        openAPIAppConfigListeners.add(listener);
+    }
+
+    /**
+     * Takes the server.xml values for ex/includedApps, ex/includedModules, and filters them down into
+     * the single comma seperated string that the existing codebase expects.
+     */
+    private void processModuleConfig(Map<String, Object> properties) {
```
This comment is no longer correct

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,455 @@
+            if (!ProductInfo.getBetaEdition()) {
+                serverxmlmode = false;
+            } else {
+                serverxmlmode = configFromServerXML.getIncludedAppsAndModules().isPresent() ||
+                                configFromServerXML.getIncludedAppsAndModules().isPresent() ||
+                                configFromServerXML.getConfigMode().isPresent();
```
```suggestion
                serverxmlmode = configFromServerXML.getIncludedAppsAndModules().isPresent() ||
                                configFromServerXML.getExcludedAppsAndModules().isPresent() ||
                                configFromServerXML.getConfigMode().isPresent();
```

Though this code may go away anyway if you replace these three methods with one method.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,455 @@
+
+    private volatile ConfigValues configValues = null;
+
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.STATIC, unbind = "unbindAppConfigListener")
+    private OpenAPIAppConfigProvider configFromServerXML;
```
```suggestion
    @Reference
    private OpenAPIAppConfigProvider configFromServerXML;
```

`cardinality=MANDATORY` and `policy=STATIC` are the defaults, they can stay if you prefer but they're not necessary.

As this reference is now a field there's no `unbind` method, so that part should definitely go.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,455 @@
+                }
+            }
+
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
+                Tr.exit(this, tc, "ConfigValues", toString());
+            }
```
I'm unsure what this does. Don't you get the `toString()` out by default in exit trace anyway? Also the `ConfigValues` object doesn't have a `toString` implementation.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,455 @@
+            }
+
+            for (ModuleName unmatchedInclude : excludedNotYetSeenButSeenUnderOldNaming.keySet()) {
+                Tr.warning(tc, MessageConstants.OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W, "excludeApplication or excludeModule", unmatchedInclude.toString(),
```
We cannot substitute "excludeApplication or excludeModule" into a translated message.

The include message further up in the file has the same issue.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,270 @@
+            //We only check equality if the server.xml is updated while the app is live
+            //And we don't want to pay the performance cost unless that happens.
+
+            includedApps.sort(Comparator.naturalOrder());
+            excludedApps.sort(Comparator.naturalOrder());
+            includedModules.sort(Comparator.naturalOrder());
+            excludedModules.sort(Comparator.naturalOrder());
+
+            otherConfig.includedApps.sort(Comparator.naturalOrder());
+            otherConfig.excludedApps.sort(Comparator.naturalOrder());
+            otherConfig.includedModules.sort(Comparator.naturalOrder());
+            otherConfig.excludedModules.sort(Comparator.naturalOrder());
```
We mustn't modify the object's state in the equals method. This could cause a concurrent modification exception if another thread is reading these fields while the sort is occurring.

We could either:
- Not sort before doing the comparison
- Copy before sorting
- Copy into a set instead of sorting
- Store the data in a set rather than a list in the first place

I'm not sure which is the best option. I'm not sure that we really need to sort before comparing since if the config is not changed in the server.xml, we should get the elements back in the same order.

On the other hand, if we really don't care about the order, storing the data in sets instead of lists makes more sense.

benjamin-confino replied:
I was thinking of one edge case where the before is 

```
<includeApplication>one</includeApplication>
<includeApplication>two</includeApplication>
```

And the after is 

```
<includeApplication>two</includeApplication>
<includeApplication>one</includeApplication>
```

Which is logically equivalent but maybe that's not worth the trouble

benjamin-confino replied:
The reason its not a set was thinking about someone doing 
```
<includeApplication>one</includeApplication>
<includeApplication>one</includeApplication>
```
I thought it was best to accurately represent that info inside java, encase we want to use it for something.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,270 @@
+            includedModulesList.ifPresent((List<String> list) -> warnInvalidModules(list, INCLUDE_MODULE_PROPERTY_NAME));
+            excludedModulesList.ifPresent((List<String> list) -> warnInvalidModules(list, EXCLUDE_MODULE_PROPERTY_NAME));
+
+            if (includedAppList.isPresent() && includedAppList.get().size() == 1) {
+                String includedString = includedAppList.get().get(0).toLowerCase().trim();
+                if (includedString.equals("all")) {
+                    configMode = Optional.of(ConfigMode.All);
+                    includedAppList = Optional.empty();
+                } else if (includedString.equals("first")) {
+                    configMode = Optional.of(ConfigMode.First);
+                    includedAppList = Optional.empty();
+                } else if (includedString.equals("none")) {
+                    configMode = Optional.of(ConfigMode.None);
+                    includedAppList = Optional.empty();
+                }
+            }
```
This logic seems questionable. Consider:

```xml
<mpOpenAPI>
  <includeApplication>all</includeApplication>
  <includeApplication>test1</includeApplication>
</mpOpenAPI>
```

With the current code, I think only `test1` would be included, and we'd treat `all` as an application name and emit a confusing warning.

Could be a follow up issue, but it needs resolved before release.

benjamin-confino replied:
Discussed in slack, we decided we'll need to think further on what's correct here.

Azquelt replied:
Still uncertain about this. I don't like that the meaning of `<includeApplication>all</includeApplication>` changes when you add another `<includeApplication>` element.

Will take another look in a follow-up issue.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,270 @@
+                openAPIAppConfigListeners.sort((OpenAPIAppConfigListener o1, OpenAPIAppConfigListener o2) -> Integer.compare(o1.getConfigListenerPriority(),
+                                                                                                                             o2.getConfigListenerPriority()));
+
+                for (OpenAPIAppConfigListener listener : openAPIAppConfigListeners) {
+                    listener.processConfigUpdate();
+                }
```
Need to check what would happen if an exception were to be thrown here.

I'm unsure whether that's possible to happen here, or how declarative services would handle it.

benjamin-confino replied:
Perhaps the question is what should happen. Lets say that DS handles it. If the first listener throws something that will prevent the second from running and, I think, leave us with the pre-updated documentation. Is that what we want?

Azquelt replied:
Hmm, I think we're ok with [the DS side](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.component.html#service.component-modified.method):
> If the modified method throws an exception, SCR must log an error message containing the exception with the Log Service, if present and continue processing the modification.

So hopefully, if there's a condition we haven't considered, the user gets an error message out that they can report.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,455 @@
+        configFromServerXMLProvider.unregisterAppConfigListener(this);
+    }
+
+    @Override
+    public String toString() {
+        if (configValues == null) {
+            return ("Unconfigured Module Selection Config");
+        }
+
+        //Set this here to avoid an infinite trace loop.
+        ConfigValues config = getConfigValues();
+        StringBuilder sb = new StringBuilder("Module Selection Config[");
+        if (config.isFirst) {
```
```suggestion
    @Override
    public String toString() {
        ConfigValues config = configValues;    
        if (config == null) {
            return ("Unconfigured Module Selection Config");
        }

        StringBuilder sb = new StringBuilder("Module Selection Config[");
        if (config.isFirst) {
```

Calling `getConfigValues()` may change the state of the object. It's unlikely since we're checking the field first, but we can just use the value from the field and avoid the chance.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/StartupWarningMessagesTest.java:
```diff
@@ -0,0 +1,129 @@
+import io.openliberty.microprofile.openapi20.fat.deployments.test1.DeploymentTestApp;
+import io.openliberty.microprofile.openapi20.fat.deployments.test1.DeploymentTestResource;
+
+@RunWith(FATRunner.class)
+//@Mode(TestMode.FULL)
+public class StartupWarningMessagesTest {
```
`@Mode` commented out

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ApplicationRegistryImpl.java:
```diff
@@ -66,13 +66,22 @@ public class ApplicationRegistryImpl implements ApplicationRegistry {
     @Reference
     private MergeProcessor mergeProcessor;
 
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.STATIC, unbind = "unbindAppConfigListener")
+    public void bindAppConfigListener(OpenAPIAppConfigProvider openAPIAppConfigProvider) {
+        openAPIAppConfigProvider.registerAppConfigListener(this);
+    }
+
+    public void unbindAppConfigListener(OpenAPIAppConfigProvider openAPIAppConfigProvider) {
+        openAPIAppConfigProvider.unregisterAppConfigListener(this);
+    }
```
I've noticed that this is a static reference.

We need to test what happens if the `<mpOpenAPI>` config element is removed altogether.

The component is set to optional configuration, but we should check whether that component ever gets deactivated since that would deactivate us and could cause lots of things to restart.

benjamin-confino replied:
I tested manually with osgiConsole and it looks like we still have OpenAPIAppConfigProvider. This is from after I removed the tag from a running server: 

```
io.openliberty.microprofile.openapi.internal.common.services.OpenAPIAppConfigProvider$OpenAPIAppConfigListener}={service.id=359, service.bundleid=130, service.scope=bundle, osgi.ds.satisfying.condition.target=(osgi.condition.id=true), component.name=io.openliberty.microprofile.openapi20.internal.ModuleSelectionConfigImpl, component.id=260}
{org.eclipse.microprofile.openapi.spi.OASFactoryResolver}={service.id=360, service.bundleid=130, service.scope=bundle, osgi.ds.satisfying.condition.target=(osgi.condition.id=true), service.vendor=IBM, component.name=io.openliberty.microprofile.openapi20.internal.OLOASFactoryResolverImpl, component.id=261}
```

```
{io.openliberty.microprofile.openapi.internal.common.services.OpenAPIAppConfigProvider}={service.id=354, service.bundleid=134, service.scope=bundle, osgi.ds.satisfying.condition.target=(osgi.condition.id=true), config.overrides=true, component.name=io.openliberty.microprofile.openapi.internal.common.OpenAPIAppConfigProviderImpl, component.id=257, config.source=file, docPath=/openapi, service.pid=io.openliberty.microprofile.openapi, config.displayId=mpOpenAPI}
```

Azquelt replied:
I'm not concerned about the `OpenAPIAppConfigProvider` being there, a lot of our test cases have no mpOpenAPI config, I'm concerned about whether it causes all the components to be restarted.

Azquelt replied:
We can note it down as something to check. Once we have a test which dynamically adds or removes the config, we can check the trace to see if the components are being deactivated and reactivated.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,270 @@
+            if (other == null) {
+                return false;
+            }
+            if (other == this) {
+                return false;
+            }
```
```suggestion
            if (other == this) {
                return true;
            }
```

benjamin-confino commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/StartupWarningMessagesTest.java:
```diff
@@ -0,0 +1,135 @@
+
+        //Example message: CWWKO1680W: The testEar application name in the includeApplication or includeModule configuration element does not match the name of any deployed application but it does match the name from the deployment descriptor of the serverXMLName application. The application name used here must be the application name specified in server.xml, or the archive file name with the ext ...
+        List<String> messages = new ArrayList<>();
+        messages.add("CWWKO1680W: The testEarIncluded/test1 application name in the includeModule configuration.*serverXMLNameIncluded");
```
Room for improvement. Replace "application name" with "module name" if possible. 

Include the module name itself in the output

Azquelt replied:
Good spot. If we're including `includeModule` or `includeApplication` here, we can probably just say "name" rather than "application name".

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,459 @@
+
+    @Override
+    public String toString() {
+        if (configValues == null) {
+            return ("Unconfigured Module Selection Config");
+        }
+
+        ConfigValues config = configValues;
+        StringBuilder sb = new StringBuilder("Module Selection Config[");
```
```suggestion
        ConfigValues config = configValues;
        if (config == null) {
            return ("Unconfigured Module Selection Config");
        }

        StringBuilder sb = new StringBuilder("Module Selection Config[");
```

Must only read the `configValues` field once since it could change. In particular it could become `null` after the initial check and cause an NPE later in the method.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,459 @@
+        return matches(name, module, getConfigValues().serverxmlmode);
+    }
+
+    private static boolean matches(ModuleName name, ModuleInfo module, boolean overridenServerXMLMode) {
```
Replace `boolean overriddenServerXMLMode` with `MatchingMode matchingMode` where `MatchingMode` is

```java
private enum MatchingMode {
    SERVER_XML_NAME,
    DEPLOYMENT_DESCRIPTOR_NAME
}
```

This should improve the readability of the callers of this method, particularly:

```java
    private boolean matches(ModuleName name, ModuleInfo module) {
        return matches(name, module, getConfigValues().serverxmlmode ? SERVER_XML_NAME : DEPLOYMENT_DESCRIPTOR_NAME);
    }
```
It's now easy to see how the whether we're in server.xml mode affects the name used for matching.

`sendWarningsForIncludesNotMatchingAnything` it should also be clearer that we're matching against both sorts of names.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,459 @@
+        return result;
+    }
+
+    /**
+     * Parses a String of app and/or module names into a list of {@code ModuleName}
+     * <p>
+     * Names must be in one of these formats:
+     * <ul>
+     * <li>appName</li>
+     * <li>appName/moduleName</li>
+     * </ul>
+     *
+     * @param nameList thane
+     * @param configKey the name of the config property holding the list (used for reporting errors)
+     * @return the parsed name or an empty Optional if it did not fit the format.
+     */
+    private static Optional<ModuleName> parseModuleName(String name, String configKey) {
```
```suggestion
    /**
     * Parses an app or module name into a {@code ModuleName}
     * <p>
     * The name must be in one of these formats:
     * <ul>
     * <li>appName</li>
     * <li>appName/moduleName</li>
     * </ul>
     *
     * @param name the name to parse
     * @param configKey the name of the config property holding the list (used for reporting errors)
     * @return the parsed name or an empty {@code Optional} if it did not fit the format.
     */
    private static Optional<ModuleName> parseModuleName(String name, String configKey) {
```

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,459 @@
+        }
+    }
+
+    private class ConfigValues {
+        protected boolean isAll = false;
+        protected boolean isFirst = false;
+        protected boolean serverxmlmode = false;
+        protected List<ModuleName> included;
+        protected List<ModuleName> excluded;
+
+        /**
+         * Configures a {@code ModuleSelectionConfig}
+         * <p>
+         * This will read and parse the {@value Constants#MERGE_INCLUDE_CONFIG} and {@value Constants#MERGE_EXCLUDE_CONFIG} config properties from mpConfig.
+         * <p>
+         * TODO doc what it reads from server.xml
+         * <p>
+         * If the config is invalid, this method will output warning messages but still return a usable result object.
+         *
+         * @return the module selection config
+         */
+        protected ConfigValues() {
```
```suggestion
    /**
     * A consistent set of configuration values used for module selection
     * <p>
     * Callers should get an instance by calling {@code getConfigValues()}.
     */
    private class ConfigValues {
        protected boolean isAll = false;
        protected boolean isFirst = false;
        protected boolean serverxmlmode = false;
        protected List<ModuleName> included;
        protected List<ModuleName> excluded;

        /**
         * Creates a new ConfigValues instance by reading the current values from server.xml and MP Config
         */
        protected ConfigValues() {
```

Update the Javadoc

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,459 @@
+                serverxmlmode = configFromServerXML.wasAnyConfigFound();
+            }
+
+            if (serverxmlmode) {
+                Tr.debug(this, tc, "Acquired includes statement from server.xml");
+
+                if (configFromServerXML.getConfigMode().isPresent()) {
+                    ConfigMode configMode = configFromServerXML.getConfigMode().get();
+                    if (configMode == OpenAPIServerXMLConfig.ConfigMode.All) {
+                        isAll = true;
+                    } else if (configMode == OpenAPIServerXMLConfig.ConfigMode.First) {
+                        isFirst = true;
+                    } else if (configMode == OpenAPIServerXMLConfig.ConfigMode.None) {
+                        included = Collections.emptyList();
+                    }
+                } else if (configFromServerXML.getIncludedAppsAndModules().isPresent()) {
+                    List<String> rawNames = configFromServerXML.getIncludedAppsAndModules().get();
+                    included = rawNames.stream().map((String rawName) -> parseModuleName(rawName, Constants.MERGE_INCLUDE_CONFIG))
+                                       .filter(Optional::isPresent)
+                                       .map(Optional::get)
+                                       .collect(Collectors.toList());
+                }
```
It's possible to make it through here with `isAll` and `isFirst` being `false` and `included` being `null` which I think will cause problems elsewhere.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java:
```diff
@@ -0,0 +1,459 @@
+    }
+
+    private class ConfigValues {
+        protected boolean isAll = false;
+        protected boolean isFirst = false;
+        protected boolean serverxmlmode = false;
+        protected List<ModuleName> included;
+        protected List<ModuleName> excluded;
```
```suggestion
        protected final boolean isAll = false;
        protected final boolean isFirst = false;
        protected final boolean serverxmlmode = false;
        protected final List<ModuleName> included;
        protected final List<ModuleName> excluded;
```

jantley-ibm commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -23,4 +23,16 @@ docPath=MicroProfile OpenAPI document endpoint path
+uiPath.desc=Specifies the URL path for accessing the MicroProfile OpenAPI UI. The URL path must be constructed with Unicode alphanumeric characters A-Za-z0-9, underscore (_), dash (-), forward slash (/), or period (.) If no value is specified, the path is set to $docPath/ui.
+
+includeApplication=MicroProfile OpenAPI included applications
+includeApplication.desc=Application names, one per element, that are to be included in the OpenAPI document. The name of an application may be defined in the server.xml, otherwise it defaults to the application archive filename excluding any extension.  The "all" special value includes all available applications. 
```
replace the word "may" with the word "can", i.e.,
The name of an application can be defined

"May" is avoided because it can confuse translators, as it can mean possibility or granting of permission.

jantley-ibm commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -23,4 +23,16 @@ docPath=MicroProfile OpenAPI document endpoint path
+excludeApplication.desc=Application names, one per element, that are to be excluded from the OpenAPI document.
+
+includeModule=MicroProfile OpenAPI included modules
+includeModule.desc=Module names, one per element, that are to be included in the OpenAPI document. Module names must be provided in the format {ApplicationName}/{ModuleName}. The name of a module may be defined in its deployment descriptor, otherwise it defaults to the module archive filename excluding any extension. It is also logged in the SRVE0169I message.
```
same as above, replace "may" with "can"

jantley-ibm reviewed (commented):
ID review comments attached; let me know if you have any questions or concerns. When we agree on changes I will add the "ID reviewed" label. Thanks --

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ApplicationRegistryImpl.java:
```diff
@@ -242,6 +249,9 @@ public OpenAPIProvider getOpenAPIProvider() {
 
             cachedProvider = result;
+            if (LoggingUtils.isEventEnabled(tc)) {
+                Tr.event(tc, "Finished creating OpenAPI provider");
```
```suggestion
                Tr.event(this, tc, "Finished creating OpenAPI provider");
```

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java:
```diff
@@ -0,0 +1,255 @@
+        openAPIAppConfigListeners.add(listener);
+    }
+
+    @Override
+    public void unregisterAppConfigListener(OpenAPIAppConfigListener listener) {
+        openAPIAppConfigListeners.add(listener);
+    }
```
```suggestion
    @Override
    public void unregisterAppConfigListener(OpenAPIAppConfigListener listener) {
        openAPIAppConfigListeners.remove(listener);
    }
```

In addition, `openAPIAppConfigListeners` needs synchronization, or to be a concurrent list

Azquelt replied:
or a copy on write list

Azquelt reviewed (dismissed):
We'll have some follow up items, but this looks good, thank you.

benjamin-confino committed:
update code for mpOpenApiElements to support new options

benjamin-confino committed:
create metatype for openapi 4.0

benjamin-confino committed:
create service interfaces for reading mpOpenAPI server xml

benjamin-confino committed:
implement service for accessing mpOpenAPI server.xml

benjamin-confino committed:
create new translation files for mpopenapi 4.0

benjamin-confino committed:
create fat tests for merging using server.xml config

benjamin-confino committed:
create fat test for warning messages for invalid mpOpenAPI server.xml config

benjamin-confino commented:
#build

benjamin-confino commented:
#build

benjamin-confino committed:
fix for failure found in testing

benjamin-confino commented:
#build

Azquelt commented:
#libby

benjamin-confino commented:
https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?uuid=76a65599-e33b-40f3-9266-96c000568fda&tab=Steps has all the test passing, as soon as the copyright header check is done we can merge

benjamin-confino commented:
#build

benjamin-confino commented:
#libby

benjamin-confino committed:
second batch of ID review changes

Azquelt commented:
#libby
#build

benjamin-confino commented:
#build

benjamin-confino commented:
#libby

Azquelt commented:
Personal builds show the following failure:

```
java.lang.Exception: Errors/warnings were found in server OpenAPIMergeWithServerXMLTestServer-_EE10_FEATURES_MicroProfile_61 logs:
 <br>[10/18/24, 18:33:28:640 UTC] 0000002a com.ibm.ws.app.manager.AppMessageHelper                      E CWWKZ0059E: The serverXMLName application installed from testEar.ear has been deleted while it is still configured.
 <br>[10/18/24, 18:33:28:659 UTC] 0000002a com.ibm.ws.app.manager.AppMessageHelper                      W CWWKZ0014W: The application serverXMLName could not be started as it could not be found at location testEar.ear.
	at componenttest.topology.impl.LibertyServer.checkLogsForErrorsAndWarnings(LibertyServer.java:3485)
	at componenttest.topology.impl.LibertyServer.stopServer(LibertyServer.java:3306)
	at componenttest.topology.impl.LibertyServer.stopServer(LibertyServer.java:3147)
	at componenttest.topology.impl.LibertyServer.stopServer(LibertyServer.java:3141)
	at componenttest.topology.impl.LibertyServer.stopServer(LibertyServer.java:3117)
	at componenttest.topology.impl.LibertyServer.stopServer(LibertyServer.java:3026)
	at componenttest.topology.impl.LibertyServer.stopServer(LibertyServer.java:3001)
	at io.openliberty.microprofile.openapi20.fat.deployments.MergeServerXMLTest.shutdownServer(MergeServerXMLTest.java:86)
```

I will see if I can fix the app deployment logic.

Based on the error message, the steps to deploy should be
- export the app archive into the apps directory
- update the server configuration
- ensure the app starts

The steps to undeploy should be
- update the server configuration
- ensure the app stops
- delete the app archive

Azquelt committed:
mpOpenAPI: Clean up test app deployment logic

Avoid deleting the app archive from /apps before we've removed its
configuration from server.xml and the app has stopped to avoid
intermittent server error messages.

Simplify the app deployment and cleanup logic. Use LibertyServer's app
start/stop validation rather than looking for log messages directly.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/MergeServerXMLTest.java:
```diff
@@ -0,0 +1,258 @@
+                                                .addIncludedApplicaiton("all")
+                                                .buildAndPushToServer();
+
+        //The combo of all and an exclude should be all except testEar/test2, but since this is set via
+        //mpConfig this should be ignored.
+        setMergeConfig(null, "testEar/test2", null);
```
This doesn't test what it's meant to because we can't actually update the system properties here while the server is running.

Noted this as something to fix in #29874

Azquelt commented:
#libby
#build

Azquelt commented:
RTC personal build is good. CI-O build is still running but has passed the relevant FATs.
